src/json-patch-duplex.ts

```diff
@@ -524,95 +540,253 @@ module jsonpatch {
             }
-            key = parseInt(key, 10);
+            key = ~~key;
           }
           if (t >= len) {
-            if (validate && patch.op === "add" && key > obj.length) {
-              throw new JsonPatchError("The specified index MUST NOT be greater than the number of elements in the array", "OPERATION_VALUE_OUT_OF_BOUNDS", p - 1, patch.path, patch);
+            if (validateOperation && operation.op === "add" && key > obj.length) {
+              throw new JsonPatchError("The specified index MUST NOT be greater than the number of elements in the array", "OPERATION_VALUE_OUT_OF_BOUNDS", 0, operation.path, operation);
+            }
+            returnValue = arrOps[operation.op].call(operation, obj, key, document); // Apply patch
+            if (returnValue.test === false) {
+              throw new JsonPatchError("Test operation failed", 'TEST_OPERATION_FAILED', 0, operation, document);
             }
-            results.push(arrOps[patch.op].call(patch, obj, key, tree)); // Apply patch
-            break;
+            return returnValue;
           }
         }
         else {
-          if (key && key.indexOf('~') != -1)
-            key = key.replace(/~1/g, '/').replace(/~0/g, '~'); // escape chars
+          if (key && key.indexOf('~') != -1) {
+            key = unescapePathComponent(key);
+          }
           if (t >= len) {
-            results.push(objOps[patch.op].call(patch, obj, key, tree)); // Apply patch
-            break;
+            returnValue = objOps[operation.op].call(operation, obj, key, document); // Apply patch
+            if (returnValue.test === false) {
+              throw new JsonPatchError("Test operation failed", 'TEST_OPERATION_FAILED', 0, operation, document);
+            }
+            return returnValue;
           }
         }
         obj = obj[key];
       }
     }
+  }
+
+  /**
+   * Apply a full JSON Patch array on a JSON document.
+   * Returns the {newDocument, result} of the patch.
+   *
+   * @param document The document to patch
+   * @param patch The patch to apply
+   * @param validateOperation `false` is without validation, `true` to use default jsonpatch's validation, or you can pass a `validateOperation` callback to be used for validation.
+   * @return An array of `{newDocument, result}` after the patch
+   */
+  export function applyPatch<T>(document: T, patch: Operation[], validateOperation?: boolean | Validator<T>): OperationResult<T>[] {
+    const results: OperationResult<any>[] = new Array(patch.length);
+
+    for (let i = 0, length = patch.length; i < length; i++) {
+      results[i] = applyOperation(document, patch[i], validateOperation);
+      document = results[i].newDocument; // in case root was replaced
+    }
+    (<any>results).newDocument = document;
```

Comment:
yeah, the cast is needed since at the time of initialising `newDocument` is not set even though it's not optional. You can also write the slightly prettier 
```ts
const results = new Array(patch.length) as <PatchResult<T>;
```
Btw, is there a perf benefit in using the array constructor? I once read that array literals are faster